modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java

```diff
@@ -0,0 +1,325 @@
+import java.nio.file.Path;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.geotools.data.shapefile.index.quadtree.QuadTree;
+import org.geotools.data.shapefile.shp.IndexFile;
+import org.geotools.data.shapefile.shp.ShapeType;
+import org.geotools.data.shapefile.shp.ShapefileReader;
+import org.geotools.data.shapefile.shp.ShapefileReader.Record;
+import org.geotools.util.NIOUtilities;
+import org.geotools.util.logging.Logging;
+import org.locationtech.jts.geom.Envelope;
+import org.locationtech.jts.index.quadtree.Quadtree;
+
+/**
+ * Utility class to assist {@link ShapeFileIndexer} in speeding up the {@link QuadTree} optimization
+ * phase through a {@link BoundsReader strategy object} that provides quick access to each shapefile
+ * record envelope, potentially avoiding an immense amount of random disk I/O calls through {@link
+ * ShapefileReader}, as the quad tree internal nodes get split/shrank.
+ *
+ * <p>Since the {@link QuadTree} leaf nodes hold only the shapefile record ids, and not their
+ * bounds, the tree layout optimization phase may incur into too much random disk reads on the
+ * {@code .shp} file.
+ *
+ * <p>The strategy object {@link #create provided} is meant to avoid that to the extent possible.
+ *
+ * <p>To a given point, record bounds will be stored in heap memory (up to 1MiB, accounting for 32K
+ * records, or 64K records if it's a points shapefile).
+ *
+ * <p>For a bigger number of {@link IndexFile#getRecordCount() shapefile records}, the strategy is
+ * to store the bounds in a temporary file (named {@code GeoTools_shp_qix_bounds_<random
+ * number>.tmp} under {@code java.io.tmpdir}), which is memory mapped and deleted at {@link
+ * BoundsReader#close()}. This leverages the Operating System's native paging, and due to the
+ * reduced size of the bounds file compared to the actual {@code.shp} and avoiding the parsing
+ * performed by {@link ShapefileReader#nextRecord()}, results in dramatically less random I/O and
+ * computing.
+ *
+ * <p>Note, however, that if there's not enough temporary space in the file system where the {@code
+ * java.io.tmpdir} directory resides, a fall back strategy that reads directly from the {@link
+ * ShapefileReader} will be used. This should be a very edge case though, since with a bounds record
+ * size of 32 bytes, the required temporary storage is 30.1MiB per million features.
+ */
+class ShapefileIndexerBoundsHelper {
+
+    private static final Logger LOGGER = Logging.getLogger(ShapefileIndexerBoundsHelper.class);
+
+    /**
+     * A record count threshold past which bounds storage falls back to a temporary file.
+     *
+     * <p>Set to the equivalent of 1MB of heap memory, accounting for 32768 records (for a 32 byte
+     * serialized envelope size), or 65536 records if the shapefile geometry type is POINT (using 16
+     * bytes per envelope).
+     */
+    private static final int FALLBACK_TO_FILE_RECCOUNT_THRESHOLD = (1024 * 1024) / 32;
```

Comment:
```suggestion
    private static final int FALLBACK_TO_FILE_REC_COUNT_THRESHOLD = (1024 * 1024) / 32;
```

(at first I told myself... is that a typo? looks like "raccoon"!)

Reply:
lol, done. Thanks for the review.